src/UiModes.ts

```diff
@@ -0,0 +1,321 @@
+                UiModeSystem.getUiModes().forEach(function(uiMode:UiMode){
+                    const uiModeObj = {
+                        name : uiMode.getName(),
+                        description : uiMode.getDescription(),
+                        settingValues : <any[]>[]
+                    }
+                    uiMode.getSettings().forEach(function(setting:SettingData){
+                        const settingObj = {
+                            key : setting.getKey(),
+                            value : setting.getValue()
+                        }
+                        uiModeObj.settingValues.push(settingObj)
+                    })
+                    uiModesObj.push(uiModeObj)
+                })
+
+                localStorage.setItem('UiModes', JSON.stringify(uiModesObj));
+                localStorage.setItem('activeUiMode', UiModeSystem.getActiveUiMode().getName());
+                UiModeSystem.localStorageUpdateCooldown = false;
+            }, 1000)
+        }else{
+            return
+        }
+    }
+
+    static loadFromLocalStorage = () : void => {
+        const uiModesObj : any[] = JSON.parse(localStorage.getItem('UiModes'))
+
+        if(uiModesObj === null){
+            return
+        }
+
+        uiModesObj.forEach(function(uiModeObj){
+            let destUiMode : UiMode = UiModeSystem.getUiModeByName(uiModeObj.name)
+            if(destUiMode===null){
+                const settings : SettingData[] = []
+                
+                Setting.getSettings().forEach(function(settingsGroup){
+                    settingsGroup.getSettings().forEach(function(setting){
+                        settings.push(new SettingData(setting.getKey(),setting.getGraphDefaultVal(),setting.getPerpetual()))
+                    })
+                })
+
+                destUiMode = new UiMode(uiModeObj.name, uiModeObj.description, settings, false)
+            }
+            uiModeObj.settingValues.forEach(function(settingObj:any){
+                destUiMode.setSettingByKey(settingObj.key,settingObj.value)
+            } )
+
+        })
+    }
+
+    static updateSettingsArray = () : void => {
+        UiModeSystem.getActiveUiMode().getSettings().forEach(function(setting){
+            Setting.setValue(setting.getKey(),setting.getValue())
+        })
+    }
+
+    static getActiveSetting = (settingName:string) : any => {
+        let found = false
+        let value:any = null
+        UiModeSystem.getActiveUiMode().getSettings().forEach(function(setting){
+            if(setting.getKey() === settingName){
+                value = setting.getValue();
+                found = true;
+            }
+        })
+        if(!found){
+            console.warn('Requested setting: "'+ settingName+'" can not be found')
+        }else{
+            return value
+        }
+    }
+
+    static setActiveSetting = (settingName:string,newValue:any) : any => {
+        let found = false
+        let settingIsPerpetual = false
```

Comment:
Instead of a boolean for settingIsPerpetual here, you could use:

let activeSetting: Setting = null;

Then inside the first forEach loop (lines 180-184) just copy the reference to the setting with the matching key:

(line 182) activeSetting = setting

Then line 186 could become:

if (activeSetting.isPerpetual())

But the big gain would be that the loop from lines 196-201 would not be required, since you already have a reference to activeSetting

You might not need the "found" boolean too, you could just check if activeSetting !== null